### DISCUSSION TITLE
Using Virtual Aperture (Region of Interest) Functionality for Lazy Signal

### DISCUSSION THREAD

**Discussion by OverEthanol** [hyperspy/hyperspy#3390] [Q&A] [answered]
Hi,

I would like to apply virtual aperture onto a 2 TB of 4DSTEM TKD data set (500, 500, 2048, 2048). I could lazy load the data set `s` by chunking it with a chunk size of (1, 10, 2048, 2048) and plot out the map using `s.plot()` just fine. But when I tried to use the `hs.plot.plot_roi_map(s)` function to apply a virtual aperture onto a region of interest (RoI), it returns with the error that 

``AttributeError: 'LazySignal' object has no attribute 'any'``

Is it because the `hs.plot.plot_roi_map(s)` has not yet been implemented for the lazy signal yet? Or is there specific argument that I must call?

If neither, how can I approach it or write my own code for plotting it? Because in my understanding, `s.plot()` given `s` is a lazy signal is just executing the instruction list of summing up the intensities of each signal pixel and using that as the pixel value for the navigator. Such "instruction/integration" method must be possible to be implemented for a selected RoI as well. 

Perhaps the bottle neck is that `hs.plot.plot_roi_map(s)` is interactive meaning as user changes the size/area of the RoI, the dask instruction list must be updated simultaneously as well and that is not feasible? which implies that to plot out the virtual image based on the virtual aperture, the virtual aperture must be predefined instead of interactively created by the user on the signal image?

Any help or suggestion would be appreciated! Thanks in advance!

**Comment by CSSFrancis** [comment, marked as answer]
Hi @OverEthanol this a quite a large dataset. Usually the live virtual images only really are possible in memory.

Maybe with great compression and feeding things through a GPU you might be able to get that to work.


That being said perhaps it is worth it to think about what you want. 2048 x 2024 is quite a few pixels.  You could `rebinned = s.rebin(scale=(1,1,16,16))` the data by a factor of 16 in the reciprocal space dimension to give you a 128x128 dataset.  That should reduce the data by a factor of 256 or the dataset should be around 8 GB which is much more manageable for these kind of live images.

You can also just make virtual images if you already know where you want to be. 

These are some examples of that.

https://pyxem.readthedocs.io/en/stable/examples/index.html#virtual-imaging


You can also think about making a segmented detector or getting the azimuthal integral in 1 dimension. You lose some information but it makes the processing simpler.
